janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphStep.java

```diff
@@ -87,14 +88,18 @@ public JanusGraphStep(final GraphStep<S, E> originalStep) {
             if (globalQuery != null && !globalQuery.getSubQuery(0).getBackendQuery().isEmpty()) {
                 globalQuery.observeWith(queryProfiler.addNested(QueryProfiler.GRAPH_CENTRIC_QUERY));
                 queries.put(0, globalQuery);
-            } else {
-                for (Map.Entry<List<HasContainer>, QueryInfo> c : hasLocalContainers.entrySet()) {
+            } else if (hasLocalContainers.size() == 1) {
+                for (Map.Entry<List<HasContainer>, QueryInfo> c : hasLocalContainers.values().stream().findFirst().get().entrySet()) {
```

Comment:
(nitpick) using `values().iterator().next()` might be slightly faster

Reply:
fair, updated